### DISCUSSION TITLE
Beartype

### DISCUSSION THREAD

**Discussion by daxpryce** [microsoft/graspologic#808] [Ideas]
This is less a feature and more a ceremony discussion, but what do you guys think about using https://github.com/beartype/beartype instead of rolling our own typechecking everywhere? 

If not beartype, we can create our own annotations and maybe write our preconditions as annotations so that we don't have to rewrite the same "nonnegative" "is a float" "is an optional str" validation steps and whatnot in every callable or constructor entry point?

**Comment by daxpryce** [comment]
@j1c @asaadeldin11 @bdpedigo @jonmclean @Nyecarr

**Comment by bdpedigo** [comment]
I'm for using something like this - I don't know anything about Beartype and how it compares to any other possible solution though. 

I am guessing it will let us do "is this an int" and not "is this an int greater than 0"?

If that is correct, then maybe we can get by with a mix of:
- Beartype (or similar) to handle "what type are you"
- sklearn.utils (in the public API) https://scikit-learn.org/stable/modules/classes.html#module-sklearn.utils to handle things like "are you greater than 0, does your input shape make sense, etc" wherever possible. For example, see https://scikit-learn.org/stable/modules/generated/sklearn.utils.check_scalar.html#sklearn.utils.check_scalar
- (if needed) write our own preconditions module, but hopefully this is a short list of things after the two above

**Reply by daxpryce** [reply]
I think that is correct, I believe the idea is:
- beartype: types are correct, no explicit typechecks -> raise TypeError in our code
- something else to handle value checks
